core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -521,7 +526,13 @@ public void halfClosed() {
           return;
         }
 
-        listener.onHalfClose();
+        GrpcCallContext.enter(callContext);
```

Comment:
I'm also not sure whether this is the right place to set & clear the ThreadLocal.  I tried doing it earlier and clearing it in closed(), but the tests showed that wasn't entirely correct.